storage/database/src/key.rs

```diff
@@ -1,54 +1,173 @@
     ///     std::cmp::Ordering::Less,
     /// );
+    ///
+    /// // Integer comparison.
+    /// let byte1 = [0, 1]; // 256
+    /// let byte2 = [255, 0]; // 255
+    /// let num1 = u16::from_le_bytes(byte1);
+    /// let num2 = u16::from_le_bytes(byte2);
+    /// assert_eq!(num1, 256);
+    /// assert_eq!(num2, 255);
     /// assert_eq!(
-    ///     <u64 as Key>::compare([1].as_slice(), [1].as_slice()),
-    ///     std::cmp::Ordering::Equal,
-    /// );
-    /// assert_eq!(
-    ///     <u64 as Key>::compare([2].as_slice(), [1].as_slice()),
+    ///     //                                               256 > 255
+    ///     <u16 as Key>::KEY_COMPARE.as_compare_fn::<u16>()(&byte1, &byte2),
     ///     std::cmp::Ordering::Greater,
     /// );
     /// ```
-    #[inline]
-    fn compare(left: &[u8], right: &[u8]) -> Ordering {
-        left.cmp(right)
-    }
+    const KEY_COMPARE: KeyCompare = KeyCompare::Lexicographic;
 }
 
 //---------------------------------------------------------------------------------------------------- Impl
-impl<T> Key for T
-where
-    T: Storable + Sized,
-{
-    type Primary = Self;
+/// [`Ord`] comparison for arrays/vectors.
+impl<const N: usize, T> Key for [T; N] where T: Key + Storable + Sized + bytemuck::Pod {}
+impl<T: bytemuck::Pod + Debug + Ord> Key for StorableVec<T> {}
+
+/// [`Ord`] comparison for misc types.
+///
+/// This is not a blanket implementation because
+/// it allows outer crates to define their own
+/// comparison functions for their `T: Storable` types.
+impl Key for () {}
+impl Key for StorableBytes {}
+impl Key for StorableStr {}
+
+/// Number comparison.
+///
+/// # Invariant
+/// This must _only_ be implemented for [`u32`], [`u64`], [`usize`].
+///
+/// This is because:
+/// 1. We use LMDB's `INTEGER_KEY`[^1] flag when this enum variant is used
+/// 2. LMDB only supports these types when using that flag[^2]
+///
+/// See: <https://docs.rs/heed/0.20.0-alpha.9/heed/struct.DatabaseFlags.html#associatedconstant.INTEGER_KEY>
+///
+/// Other numbers will still have the same behavior, but they use
+/// [`impl_custom_numbers_key`] and essentially pass LMDB a "custom"
+/// number compare function.
+macro_rules! impl_number_key {
+    ($($t:ident),* $(,)?) => {
+        $(
+            impl Key for $t {
+                const KEY_COMPARE: KeyCompare = KeyCompare::Number;
+            }
+        )*
+    };
+}
+
+impl_number_key!(u32, u64, usize);
+
+/// Custom number comparison for other numbers.
+macro_rules! impl_custom_numbers_key {
+    ($($t:ident),* $(,)?) => {
+        $(
+            impl Key for $t {
+                // Just forward the the number comparison function.
+                const KEY_COMPARE: KeyCompare = KeyCompare::Custom(|left, right| {
+                    KeyCompare::Number.as_compare_fn::<$t>()(left, right)
+                });
+            }
+        )*
+    };
+}
+
+impl_custom_numbers_key!(u8, u16, u128, i8, i16, i32, i64, i128, isize);
```

Comment:
## Key 3/4
Implementation detail: LMDB's `INTEGER_KEY` flag only supports `u32`, `u64` and `usize` if it's either of those.

`redb` supports all numbers out-the-box though, so our workaround is:

1. We implement `Key` for `u32, u64, usize` and set `KeyCompare::Number`
2. For all other numbers we set `KeyCompare::Custom(/* number compare fn */)`

then in the `heed` backend implementation...